File: trunk/+tom/DataContainerDefault.m
classdef DataContainerDefault < tom.DataContainer
  
  methods (Static = true, Access = public)
    function initialize(name)
      function text = componentDescription
        text = 'This default data container no data.';
      end
      tom.DataContainer.connect(name, @componentDescription, @tom.DataContainerDefault);
    end
  end
  
  methods (Access = public, Static = true)
    function this = DataContainerDefault(initialTime)
      this = this@tom.DataContainer(initialTime);
    end
  end
  
  methods (Access = public, Static = false)
    function list = listSensors(this, type)
      assert(isa(this, 'tom.DataContainer'));
      assert(isa(type, 'char'));
      list = repmat(tom.SensorIndex(0), [0, 1]);
    end
    
    function text = getSensorDescription(this, id)
      assert(isa(this, 'tom.DataContainer'));
      assert(isa(id, 'tom.SensorIndex'));
      text = '';
      assert(isa(text, 'char'));
      error('This default data container has no data.');
    end
        
    function obj = getSensor(this, id)
      assert(isa(this, 'tom.DataContainer'));
      assert(isa(id, 'tom.SensorIndex'));
      obj = tom.Measure.create('tom', tom.WorldTime(0), '');
      assert(isa(obj,'tom.Sensor'));
      error('This default data container has no data.');
    end
    
    function flag = hasReferenceTrajectory(this)
      assert(isa(this, 'tom.DataContainer'));
      flag = false;
    end
    
    function x = getReferenceTrajectory(this)
      assert(isa(this, 'tom.DataContainer'));
      x = tom.DynamicModel.create('tom', tom.WorldTime(0), '');
      assert(isa(x, 'tom.Trajectory'));
      error('This default data container has no data.');
    end
  end
  
end

File: trunk/+tom/DataContainerDefault.cpp
namespace tom
{
  class DataContainerDefault : public DataContainer
  {
  public:
    DataContainerDefault(const WorldTime initialTime) : DataContainer(initialTime)
    {
      return;
    }

    void listSensors(const std::string type, std::vector<SensorIndex>& list)
    {
      list.resize(0);
      return;
    }

    std::string getSensorDescription(SensorIndex id)
    {
       static std::string text("");
       throw("This default data container has no data.");
       return (text);
    }

    Sensor& getSensor(SensorIndex id)
    {
      Sensor* sensor = NULL;
      throw("This default data container has no data.");
      return (*sensor); // would crash if we ever got here
    }

    bool hasReferenceTrajectory(void)
    {
      return (false);
    }

    Trajectory* getReferenceTrajectory(void)
    {
      Trajectory* trajectory = NULL;
      throw("This default data container has no data.");
      return (trajectory);
    }

  private:
    static std::string componentDescription(void)
    {
      return ("This default data container has no data.");
    }

    static DataContainer* componentFactory(const WorldTime initialTime)
    {
      return (new DataContainerDefault(initialTime));
    }

  protected:
    static void initialize(std::string name)
    {
      connect(name, componentDescription, componentFactory);
    }
    friend class DataContainerDefaultInitializer;
  };

  class DataContainerDefaultInitializer
  {
  public:
    DataContainerDefaultInitializer(void)
    {
      DataContainerDefault::initialize("tom");
    }
  } _DataContainerDefaultInitializer;
}
